library/ssl_cli.c

```diff
@@ -1508,37 +1838,67 @@ static int ssl_parse_server_hello( mbedtls_ssl_context *ssl )
         }
 #endif /* MBEDTLS_SSL_RENEGOTIATION */
 
-        MBEDTLS_SSL_DEBUG_MSG( 1, ( "bad server hello message" ) );
-        mbedtls_ssl_send_alert_message( ssl, MBEDTLS_SSL_ALERT_LEVEL_FATAL,
-                                        MBEDTLS_SSL_ALERT_MSG_UNEXPECTED_MESSAGE );
+        MBEDTLS_SSL_DEBUG_MSG( 1, ( "bad server hello message - bad record type" ) );
+
+        ssl->send_alert = MBEDTLS_SSL_ALERT_LEVEL_FATAL;
+        ssl->alert_type = MBEDTLS_SSL_ALERT_MSG_UNEXPECTED_MESSAGE;
         return( MBEDTLS_ERR_SSL_UNEXPECTED_MESSAGE );
     }
 
+    /* In DTLS, we tolerate a HelloVerifyRequest in place of a ServerHello. */
+
 #if defined(MBEDTLS_SSL_PROTO_DTLS)
     if( ssl->conf->transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM )
     {
-        if( buf[0] == MBEDTLS_SSL_HS_HELLO_VERIFY_REQUEST )
+        if( ssl->in_msg[0] == MBEDTLS_SSL_HS_HELLO_VERIFY_REQUEST )
         {
             MBEDTLS_SSL_DEBUG_MSG( 2, ( "received hello verify request" ) );
-            MBEDTLS_SSL_DEBUG_MSG( 2, ( "<= parse server hello" ) );
-            return( ssl_parse_hello_verify_request( ssl ) );
+            ret = SSL_SERVER_HELLO_COORDINATE_VERIFY;
         }
         else
         {
             /* We made it through the verification process */
             mbedtls_free( ssl->handshake->verify_cookie );
             ssl->handshake->verify_cookie = NULL;
             ssl->handshake->verify_cookie_len = 0;
+
+            ret = SSL_SERVER_HELLO_COORDINATE_HELLO;
         }
     }
+    else
 #endif /* MBEDTLS_SSL_PROTO_DTLS */
+    {
+        ret = SSL_SERVER_HELLO_COORDINATE_HELLO;
+    }
+
+    MBEDTLS_SSL_DEBUG_MSG( 2, ( "<= process server hello - coordinate" ) );
+    return( ret );
+}
+
+static int ssl_process_server_hello_parse( mbedtls_ssl_context *ssl,
+                                           const unsigned char* buf,
+                                           size_t buflen )
+{
+    int ret, i;
+    size_t n;
+    size_t ext_len;
+    unsigned char const *ext;
+    unsigned char comp;
+#if defined(MBEDTLS_ZLIB_SUPPORT)
+    int accept_comp;
+#endif
+    const mbedtls_ssl_ciphersuite_t *suite_info;
+
+    MBEDTLS_SSL_DEBUG_MSG( 2, ( "=> process server hello - parse" ) );
 
-    if( ssl->in_hslen < 38 + mbedtls_ssl_hs_hdr_len( ssl ) ||
+    if( buflen < 38 + mbedtls_ssl_hs_hdr_len( ssl ) ||
```

Comment:
Magic number

Reply:
@AndrzejKurek I agree that this should be changed on the long run, but I prefer not to touch the actual parsing/writing routines for now more than we have to, and leave improving, and potentially adapting them to use the readers/writers, for later. This is one shortcut we can take to get to a deliverable based on the new messaging stack quicker.